File: leetcode/easy/67.Add Binary/python/addBinary.py

# 1. cast a and b to decimal numbers
# 2. sum each other
# 3. cast again to binary
# 4. return it
def add_binary(a, b):
    dec_a = 0
    dec_b = 0
    count = 0
    for i in range(len(a) - 1, -1, -1):
        if int(a[i]) == 1:
            dec_a += 2 ** count
        count += 1
    count = 0
    for i in range(len(b) - 1, -1, -1):
        if int(b[i]) == 1:
            dec_b += 2 ** count
        count += 1
    result = dec_a + dec_b
    if result == 0:
        return '0'
    bin_num_list = []
    while result >= 1:
        if result % 2 == 0:
            bin_num_list.insert(0, '0')
        else: 
            bin_num_list.insert(0, '1')
        result //= 2
    bin_num = ''
    for bi in bin_num_list:
        bin_num += bi
    return bin_num

test_a = '1010'
test_b = '1011'
print(add_binary(test_a, test_b))


# More pythonistic
def add_binary1(a, b):
    dec_a, dec_b, count = 0, 0, 0
    for i in range(len(a) - 1, -1, -1):
        if a[i] == '1':
            dec_a += 2 ** count
        count += 1
    count = 0
    for i in range(len(b) - 1, -1, -1):
        if b[i] == '1':
            dec_b += 2 ** count
        count += 1
    bin_num = bin(dec_a + dec_b)
    str_bin = bin_num[2:]
    return str_bin
    

print(add_binary1(test_a,test_b))

File: leetcode/easy/67.Add Binary/js/addBinary.js
"use strict";

/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */

var addBinary = function(a, b) {
    let decA = BigInt(`0b${a}`);
    let decB = BigInt(`0b${b}`);
    
    console.log(`dec a: ${decA}, dec b: ${decB}`);
    
    let result = decA + decB;
    console.log(result);
    console.log(typeof result);

    if (result === 0n) {
        return '0';
    }
    let str_bin = result.toString(2);
    return str_bin;
};


const test_a = "10100000100100110110010000010101111011011001101110111111111101000000101111001110001111100001101"
const test_b = "110101001011101110001111100110001010100001101011101010000011011011001011101111001100000011011110011";
console.log(addBinary(test_a, test_b));
